algorithmic_efficiency/workloads/ogb/ogb_jax/workload.py

```diff
@@ -0,0 +1,211 @@
+    rng, params_rng, dropout_rng = jax.random.split(rng, 3)
+    init_fn = jax.jit(functools.partial(self._model.init, train=False))
+    params = init_fn({'params': params_rng, 'dropout': dropout_rng},
+                     self._init_graphs)
+    params = params['params']
+    self._param_shapes = jax.tree_map(lambda x: spec.ShapeTuple(x.shape),
+                                      params)
+    return jax_utils.replicate(params), None
+
+  # Keep this separate from the loss function in order to support optimizers
+  # that use the logits.
+  def output_activation_fn(self,
+                           logits_batch: spec.Tensor,
+                           loss_type: spec.LossType) -> spec.Tensor:
+    pass
+
+  @property
+  def loss_type(self):
+    return spec.LossType.SOFTMAX_CROSS_ENTROPY
+
+  def model_fn(
+      self,
+      params: spec.ParameterContainer,
+      augmented_and_preprocessed_input_batch: spec.Tensor,
+      model_state: spec.ModelAuxiliaryState,
+      mode: spec.ForwardPassMode,
+      rng: spec.RandomState,
+      update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
+    """Get predicted logits from the network for input graphs."""
+    del update_batch_norm  # No BN in the GNN model.
+    assert model_state is None
+    train = mode == spec.ForwardPassMode.TRAIN
+    logits = self._model.apply({'params': params},
+                               augmented_and_preprocessed_input_batch,
+                               rngs={'dropout': rng},
+                               train=train)
+    return logits, None
+
+  def _binary_cross_entropy_with_mask(self,
+                                      labels: jnp.ndarray,
+                                      logits: jnp.ndarray,
+                                      mask: jnp.ndarray) -> jnp.ndarray:
+    """Binary cross entropy loss for logits, with masked elements."""
+    assert logits.shape == labels.shape == mask.shape
+    assert len(logits.shape) == 2
+
+    # To prevent propagation of NaNs during grad().
+    # We mask over the loss for invalid targets later.
+    labels = jnp.where(mask, labels, -1)
+
+    # Numerically stable implementation of BCE loss.
+    # This mimics TensorFlow's tf.nn.sigmoid_cross_entropy_with_logits().
+    positive_logits = (logits >= 0)
+    relu_logits = jnp.where(positive_logits, logits, 0)
+    abs_logits = jnp.where(positive_logits, logits, -logits)
+    return relu_logits - (logits * labels) + (jnp.log(1 + jnp.exp(-abs_logits)))
+
+  # Does NOT apply regularization, which is left to the submitter to do in
+  # `update_params`.
+  def loss_fn(
+      self,
+      label_batch: spec.Tensor,
+      logits_batch: spec.Tensor,
+      mask_batch: Optional[spec.Tensor]) -> spec.Tensor:  # differentiable
+    loss = self._binary_cross_entropy_with_mask(
+        labels=label_batch, logits=logits_batch, mask=mask_batch)
+    mean_loss = jnp.sum(jnp.where(mask_batch, loss, 0)) / jnp.sum(mask_batch)
```

Comment:
Shouldn't the loss function return the per-example instead of the mean loss?

Reply:
Yep! Thanks for catching that. Done.